src/devices/cpu/drcbearm64.cpp

```diff
@@ -0,0 +1,3876 @@
+	{
+		mov_param_imm(a, inst.size(), dstp, 1);
+		return;
+	}
+	else
+	{
+		const a64::Gp dst = dstp.select_register(TEMP_REG1, inst.size());
+
+		if (inst.condition() == COND_C || inst.condition() == COND_NC || inst.condition() == COND_A || inst.condition() == COND_BE)
+			load_carry(a, true);
+
+		if (inst.condition() == uml::COND_U || inst.condition() == uml::COND_NU)
+		{
+			get_unordered(a, dst);
+
+			if (inst.condition() == uml::COND_NU)
+				a.eor(dst, dst, 1);
+		}
+		else
+			a.cset(dst, ARM_CONDITION(a, inst.condition()));
+
+		mov_param_reg(a, inst.size(), dstp, dst);
+	}
+}
+
+void drcbe_arm64::op_mov(a64::Assembler &a, const uml::instruction &inst)
+{
+	assert(inst.size() == 4 || inst.size() == 8);
+	assert_any_condition(inst);
+	assert_no_flags(inst);
+
+	be_parameter dstp(*this, inst.param(0), PTYPE_MR);
+	be_parameter srcp(*this, inst.param(1), PTYPE_MRI);
+
+	// add a conditional branch unless a conditional move is possible
+	Label skip = a.newLabel();
+
+	if (inst.condition() != uml::COND_ALWAYS)
+	{
+		if (inst.condition() == COND_C || inst.condition() == COND_NC || inst.condition() == COND_A || inst.condition() == COND_BE)
+			load_carry(a, true);
+
+		if (inst.condition() == uml::COND_U || inst.condition() == uml::COND_NU)
+			check_unordered_condition(a, inst.condition(), skip, true);
+		else
+			a.b(ARM_NOT_CONDITION(a, inst.condition()), skip);
+	}
+
+	mov_param_param(a, inst.size(), dstp, srcp);
+
+	if (inst.condition() != uml::COND_ALWAYS)
+		a.bind(skip);
+}
+
+void drcbe_arm64::op_sext(a64::Assembler &a, const uml::instruction &inst)
+{
+	assert(inst.size() == 4 || inst.size() == 8);
+	assert_no_condition(inst);
+	assert_flags(inst, FLAG_S | FLAG_Z);
+
+	be_parameter dstp(*this, inst.param(0), PTYPE_MR);
+	be_parameter srcp(*this, inst.param(1), PTYPE_MRI);
+	const parameter &sizep = inst.param(2);
+	assert(sizep.is_size());
+	const auto size = sizep.size();
+
+	const a64::Gp dstreg = dstp.select_register(TEMP_REG2, inst.size());
+
+	if ((1 << size) >= inst.size())
+	{
+		mov_param_param(a, inst.size(), dstp, srcp);
+	}
+	else
+	{
+		if (srcp.is_memory())
+		{
+			if (size == SIZE_BYTE)
+				emit_ldrsb_mem(a, dstreg.x(), srcp.memory());
+			else if (size == SIZE_WORD)
+				emit_ldrsh_mem(a, dstreg.x(), srcp.memory());
+			else if (size == SIZE_DWORD)
+				emit_ldrsw_mem(a, dstreg.x(), srcp.memory());
+			else if (size == SIZE_QWORD)
+				emit_ldr_mem(a, dstreg.x(), srcp.memory());
+		}
+		else
+		{
+			const a64::Gp tempreg = srcp.select_register(dstreg, 8);
+			mov_reg_param(a, inst.size(), tempreg, srcp);
+
+			if (size == SIZE_BYTE)
+				a.sxtb(dstreg.x(), tempreg.w());
+			else if (size == SIZE_WORD)
+				a.sxth(dstreg.x(), tempreg.w());
+			else if (size == SIZE_DWORD)
+				a.sxtw(dstreg.x(), tempreg.w());
+		}
+
+		mov_param_reg(a, inst.size(), dstp, dstreg);
+	}
+
+	if (inst.flags())
+		a.tst(dstreg, dstreg);
+}
+
+void drcbe_arm64::op_roland(a64::Assembler &a, const uml::instruction &inst)
+{
```

Comment:
Is it worth detecting cases where this can be special-cased to use a `ubfx` or `ubfiz` instruction?  If the mask and shift are immediate:

* `ubfx` can be used if the mask is right-aligned (i.e. starts at bit position zero)
* `ubfiz` can be used if the mask is contiguous and starts at the bit position equal to the shift value

For example:
* `ROLAND(d, s, 3, 0x003f)` is equivalent to `ubfx(d, s, bits - 3, 6)`
* `ROLAND(d, s, 5, 0x07e0)` is equivalent to `ubfiz(d, s, 5, 6)`

Reply:
Real-world PowerPC RLW* usage will probably fall into one of those two cases a majority of the time, and that's not even counting the hard-coded instances the DRC generates for instructions like MFCR.

I'm not sure how far into the weeds we want to go with optimizations before landing the first working version, but this seems safe and useful.

Reply:
I don't think you can optimize ROLAND any more than it is now. You can't simplify the command down to just one ubfx/ubfiz because you need to rotate the value.

Here's what the current implementation does:
- If the mask is 0 then the value doesn't change
- If the shift is a constant then it gets simplified to a ror
- The final command is always just a single AND of the rotated value